pkg/controller/job/job_controller.go

```diff
@@ -0,0 +1,452 @@
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package job
+
+import (
+	"reflect"
+	"sort"
+	"sync"
+	"time"
+
+	"github.com/golang/glog"
+	"k8s.io/kubernetes/pkg/api"
+	"k8s.io/kubernetes/pkg/apis/experimental"
+	"k8s.io/kubernetes/pkg/client/cache"
+	"k8s.io/kubernetes/pkg/client/record"
+	client "k8s.io/kubernetes/pkg/client/unversioned"
+	"k8s.io/kubernetes/pkg/controller"
+	"k8s.io/kubernetes/pkg/controller/framework"
+	"k8s.io/kubernetes/pkg/controller/replication"
+	"k8s.io/kubernetes/pkg/fields"
+	"k8s.io/kubernetes/pkg/labels"
+	"k8s.io/kubernetes/pkg/runtime"
+	"k8s.io/kubernetes/pkg/util"
+	"k8s.io/kubernetes/pkg/util/workqueue"
+	"k8s.io/kubernetes/pkg/watch"
+)
+
+type JobManager struct {
+	kubeClient client.Interface
+	podControl controller.PodControlInterface
+
+	// To allow injection of updateJob for testing.
+	updateHandler func(job *experimental.Job) error
+	syncHandler   func(jobKey string) error
+	// podStoreSynced returns true if the pod store has been synced at least once.
+	// Added as a member to the struct to allow injection for testing.
+	podStoreSynced func() bool
+
+	// A TTLCache of pod creates/deletes each rc expects to see
+	expectations controller.ControllerExpectationsInterface
+
+	// A store of job, populated by the jobController
+	jobStore cache.StoreToJobLister
+	// Watches changes to all jobs
+	jobController *framework.Controller
+
+	// A store of pods, populated by the podController
+	podStore cache.StoreToPodLister
+	// Watches changes to all pods
+	podController *framework.Controller
+
+	// Jobs that need to be updated
+	queue *workqueue.Type
+}
+
+func NewJobManager(kubeClient client.Interface) *JobManager {
+	eventBroadcaster := record.NewBroadcaster()
+	eventBroadcaster.StartLogging(glog.Infof)
+	eventBroadcaster.StartRecordingToSink(kubeClient.Events(""))
+
+	jm := &JobManager{
+		kubeClient: kubeClient,
+		podControl: controller.RealPodControl{
+			KubeClient: kubeClient,
+			Recorder:   eventBroadcaster.NewRecorder(api.EventSource{Component: "job"}),
+		},
+		expectations: controller.NewControllerExpectations(),
+		queue:        workqueue.New(),
+	}
+
+	jm.jobStore.Store, jm.jobController = framework.NewInformer(
+		&cache.ListWatch{
+			ListFunc: func() (runtime.Object, error) {
+				return jm.kubeClient.Experimental().Jobs(api.NamespaceAll).List(labels.Everything(), fields.Everything())
+			},
+			WatchFunc: func(rv string) (watch.Interface, error) {
+				return jm.kubeClient.Experimental().Jobs(api.NamespaceAll).Watch(labels.Everything(), fields.Everything(), rv)
+			},
+		},
+		&experimental.Job{},
+		replicationcontroller.FullControllerResyncPeriod,
+		framework.ResourceEventHandlerFuncs{
+			AddFunc: jm.enqueueController,
+			UpdateFunc: func(old, cur interface{}) {
+				job := cur.(*experimental.Job)
+				for _, c := range job.Status.Conditions {
```

Comment:
In a follow-up PR, consider a comment explaining why completed job changes are ignored.